RATIS-1402. do not send extra rpc calls to follower when the follower is still installing a snapshot

JacksonYao287 opened the pull request:
## What changes were proposed in this pull request?
when notifying install snapshot, we should wait until the snapshot is installed successfully or exceptionally and then start appending subsequent log entries. 
in current implement , `wait` is used to achieve this purpose. but `notifyInstallSnapshotFromLeader` in state machine is call in a separate thread (completeFuture.supplyAsync), so `wait` does not take effect at this situation.
as a result , the client will get a reply with`installSnapshotResult.IN_PROGRESS` before the installation process is completed.
and then, then client will retry this call, because `setAttemptedToInstallSnapshot()` is not called when  received a reply with`installSnapshotResult.IN_PROGRESS`.

this will leading to too many useless rpc calls (shown in the attachment of this jira).

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-1402

## How was this patch tested?

current unit test

JacksonYao287 commented:
@szetszwo PTAL!  i have tested this patch, scm bootstraps as expected now

szetszwo commented:
@JacksonYao287 TestInstallSnapshotNotificationWithGrpc failed.  It seems related.  Please take a look.

BTW, I just have requested @hanishakoneru  to review this.

JacksonYao287 commented:
@szetszwo @hanishakoneru PTAL! thanks. i thinks this is the reason why the leader will make so many rpc calls as shown in the attachment of this jira , i have tested this patch, it works as expected. 

the test seems faild, but i do not think it is caused by this patch. i will try to fix the test

hanishakoneru commented:
@JacksonYao287 thanks for working on this.

If I understand the motivation for this fix correctly, it is to avoid the leader sending extra rpc calls to follower when the follower is still installing a snapshot. Is this correct? If yes, could you please update the Jira title to reflect that?

Using join(), we are giving up calling the StateMachine asynchronously as we wait for it to finish. Also, if the StateMachine#notifyInstallSnapshotFromLeader() completes exceptionally, join() would throw an unchecked exception which might cause the follower to shutdown.

So instead of join(), would it be better to call wait of the main thread itself for a specified time? This way, if the StateMachine completes the future before the wait if over, we can return the result. Otherwise, we can return IN_PROGRESS notification back to the leader.

JacksonYao287 commented:
thanks @hanishakoneru for the review !

>If I understand the motivation for this fix correctly, it is to avoid the leader sending extra rpc calls to follower when the follower is still installing a snapshot. Is this correct? If yes, could you please update the Jira title to reflect that?

yes it is . when i test, i found two problems。
1 leader will send a lot of extra rpc calls to follower when the follower is still installing a snapshot.
2 after snapshot is installed successfully , leader and follower can not work as expected. `nextIndex: updateUnconditionally` will be called infinitely and the new scm can not be bootstraped successfully. please take a look at the attachement of this jira

leader.log 
```
InstallSnapshotResponseHandler: received a reply b6424ca2-783b-4e68-b12c-95a0e99bb745<-b1b06ffa-12b9-4093-a844-58467d6298b9#0:FAIL-t2,SNAPSHOT_INSTALLED,snapshotIndex=2
2021-09-10 14:31:35,849 [grpc-default-executor-0] INFO org.apache.ratis.grpc.server.GrpcLogAppender: b6424ca2-783b-4e68-b12c-95a0e99bb745@group-194BB21401F4->b1b06ffa-12b9-4093-a844-58467d6298b9-InstallSnapshotResponseHandler: Follower installed snapshot at index 2
2021-09-10 14:31:35,849 [grpc-default-executor-0] INFO org.apache.ratis.server.leader.FollowerInfo: b6424ca2-783b-4e68-b12c-95a0e99bb745@group-194BB21401F4->b1b06ffa-12b9-4093-a844-58467d6298b9: snapshotIndex: setUnconditionally 0 -> 2
2021-09-10 14:31:35,849 [grpc-default-executor-0] INFO org.apache.ratis.server.leader.FollowerInfo: b6424ca2-783b-4e68-b12c-95a0e99bb745@group-194BB21401F4->b1b06ffa-12b9-4093-a844-58467d6298b9: matchIndex: setUnconditionally 0 -> 2
2021-09-10 14:31:35,849 [grpc-default-executor-0] INFO org.apache.ratis.server.leader.FollowerInfo: b6424ca2-783b-4e68-b12c-95a0e99bb745@group-194BB21401F4->b1b06ffa-12b9-4093-a844-58467d6298b9: nextIndex: setUnconditionally 3 -> 3
2021-09-10 14:31:35,849 [grpc-default-executor-0] INFO org.apache.ratis.server.leader.FollowerInfo: Follower b6424ca2-783b-4e68-b12c-95a0e99bb745@group-194BB21401F4->b1b06ffa-12b9-4093-a844-58467d6298b9 acknowledged installing snapshot
2021-09-10 14:31:35,849 [grpc-default-executor-0] INFO org.apache.ratis.server.RaftServer$Division: commitIndexChanged notify senders,org.apache.ratis.server.impl.LeaderStateImpl$SenderList@328cc3cb
2021-09-10 14:31:35,849 [grpc-default-executor-0] INFO org.apache.ratis.server.leader.FollowerInfo: b6424ca2-783b-4e68-b12c-95a0e99bb745@group-194BB21401F4->b1b06ffa-12b9-4093-a844-58467d6298b9: nextIndex: updateToMax old=3, new=3, updated? false
2021-09-10 14:31:35,849 [grpc-default-executor-0] INFO org.apache.ratis.server.leader.FollowerInfo: b6424ca2-783b-4e68-b12c-95a0e99bb745@group-194BB21401F4->b1b06ffa-12b9-4093-a844-58467d6298b9: nextIndex: updateUnconditionally 3 -> 3
2021-09-10 14:31:38,352 [grpc-default-executor-0] INFO org.apache.ratis.server.leader.FollowerInfo: b6424ca2-783b-4e68-b12c-95a0e99bb745@group-194BB21401F4->b1b06ffa-12b9-4093-a844-58467d6298b9: nextIndex: updateUnconditionally 3 -> 3

2021-09-10 14:31:40,414 [grpc-default-executor-0] INFO org.apache.ratis.server.leader.FollowerInfo: b6424ca2-783b-4e68-b12c-95a0e99bb745@group-194BB21401F4->b1b06ffa-12b9-4093-a844-58467d6298b9: nextIndex: updateUnconditionally 4 -> 3
2021-09-10 14:31:40,415 [grpc-default-executor-0] INFO org.apache.ratis.server.leader.FollowerInfo: b6424ca2-783b-4e68-b12c-95a0e99bb745@group-194BB21401F4->b1b06ffa-12b9-4093-a844-58467d6298b9: nextIndex: updateUnconditionally 4 -> 3
2021-09-10 14:31:40,415 [grpc-default-executor-0] INFO org.apache.ratis.server.leader.FollowerInfo: b6424ca2-783b-4e68-b12c-95a0e99bb745@group-194BB21401F4->b1b06ffa-12b9-4093-a844-58467d6298b9: nextIndex: updateUnconditionally 4 -> 3
2021-09-10 14:31:40,416 [grpc-default-executor-0] INFO org.apache.ratis.server.leader.FollowerInfo: b6424ca2-783b-4e68-b12c-95a0e99bb745@group-194BB21401F4->b1b06ffa-12b9-4093-a844-58467d6298b9: nextIndex: updateUnconditionally 4 -> 3

2021-09-10 14:31:40,420 [grpc-default-executor-0] INFO org.apache.ratis.server.leader.FollowerInfo: b6424ca2-783b-4e68-b12c-95a0e99bb745@group-194BB21401F4->b1b06ffa-12b9-4093-a844-58467d6298b9: nextIndex: updateUnconditionally 4 -> 3
```
follower.log
```
2021-09-10 14:31:35,856 [grpc-default-executor-0] INFO org.apache.ratis.server.RaftServer$Division: b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4: Failed appendEntries as snapshot ((t:2, i:2)) installation is in progress
2021-09-10 14:31:35,856 [grpc-default-executor-0] INFO org.apache.ratis.server.RaftServer$Division: b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4: inconsistency entries. Reply:b6424ca2-783b-4e68-b12c-95a0e99bb745<-b1b06ffa-12b9-4093-a844-58467d6298b9#207:FAIL-t2,INCONSISTENCY,nextIndex=3,followerCommit=2
2021-09-10 14:31:36,020 [b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4-StateMachineUpdater] INFO org.apache.hadoop.hdds.scm.ha.SCMHAManagerImpl: Installing checkpoint with SCMTransactionInfo 2#2
2021-09-10 14:31:36,023 [b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4-StateMachineUpdater] INFO org.apache.hadoop.hdds.scm.ha.SCMHAManagerImpl: Replaced DB with checkpoint, term: 2, index: 2
2021-09-10 14:31:36,171 [b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4-StateMachineUpdater] INFO org.apache.hadoop.hdds.scm.ha.SequenceIdGenerator: reinitialize SequenceIdGenerator.
2021-09-10 14:31:36,172 [b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4-StateMachineUpdater] INFO org.apache.hadoop.hdds.scm.pipeline.PipelineStateManager: No pipeline exists in current db
2021-09-10 14:31:36,172 [b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4-StateMachineUpdater] INFO org.apache.hadoop.hdds.scm.ha.SCMHAManagerImpl: Reloaded SCM state with Term: 2 and Index: 2
2021-09-10 14:31:36,174 [b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4-StateMachineUpdater] INFO org.apache.ratis.server.impl.StateMachineUpdater: b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4-StateMachineUpdater: snapshotIndex: setUnconditionally -1 -> 2
2021-09-10 14:31:36,174 [b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4-StateMachineUpdater] INFO org.apache.ratis.server.impl.StateMachineUpdater: b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4-StateMachineUpdater: appliedIndex: setUnconditionally -1 -> 2
2021-09-10 14:31:38,357 [grpc-default-executor-0] INFO org.apache.ratis.server.RaftServer$Division: b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4: Failed appendEntries as snapshot ((t:2, i:2)) installation is in progress
2021-09-10 14:31:38,357 [grpc-default-executor-0] INFO org.apache.ratis.server.RaftServer$Division: b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4: inconsistency entries. Reply:b6424ca2-783b-4e68-b12c-95a0e99bb745<-b1b06ffa-12b9-4093-a844-58467d6298b9#208:FAIL-t2,INCONSISTENCY,nextIndex=3,followerCommit=2
2021-09-10 14:31:40,418 [grpc-default-executor-0] INFO org.apache.ratis.server.RaftServer$Division: b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4: Failed appendEntries as snapshot ((t:2, i:2)) installation is in progress
2021-09-10 14:31:40,418 [grpc-default-executor-0] INFO org.apache.ratis.server.RaftServer$Division: b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4: inconsistency entries. Reply:b6424ca2-783b-4e68-b12c-95a0e99bb745<-b1b06ffa-12b9-4093-a844-58467d6298b9#209:FAIL-t2,INCONSISTENCY,nextIndex=3,followerCommit=2
2021-09-10 14:31:40,419 [grpc-default-executor-0] INFO org.apache.ratis.server.RaftServer$Division: b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4: Failed appendEntries as snapshot ((t:2, i:2)) installation is in progress
2021-09-10 14:31:40,419 [grpc-default-executor-0] INFO org.apache.ratis.server.RaftServer$Division: b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4: inconsistency entries. Reply:b6424ca2-783b-4e68-b12c-95a0e99bb745<-b1b06ffa-12b9-4093-a844-58467d6298b9#210:FAIL-t2,INCONSISTENCY,nextIndex=3,followerCommit=2
2021-09-10 14:31:40,420 [grpc-default-executor-0] INFO org.apache.ratis.server.RaftServer$Division: b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4: Failed appendEntries as snapshot ((t:2, i:2)) installation is in progress
2021-09-10 14:31:40,420 [grpc-default-executor-0] INFO org.apache.ratis.server.RaftServer$Division: b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4: inconsistency entries. Reply:b6424ca2-783b-4e68-b12c-95a0e99bb745<-b1b06ffa-12b9-4093-a844-58467d6298b9#211:FAIL-t2,INCONSISTENCY,nextIndex=3,followerCommit=2
2021-09-10 14:31:40,420 [grpc-default-executor-0] INFO org.apache.ratis.server.RaftServer$Division: b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4: Failed appendEntries as snapshot ((t:2, i:2)) installation is in progress
2021-09-10 14:31:40,420 [grpc-default-executor-0] INFO org.apache.ratis.server.RaftServer$Division: b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4: inconsistency entries. Reply:b6424ca2-783b-4e68-b12c-95a0e99bb745<-b1b06ffa-12b9-4093-a844-58467d6298b9#212:FAIL-t2,INCONSISTENCY,nextIndex=3,followerCommit=2
```
from follower.log, we can see  that , even if the snapshot has been installed successfully , the follower also consider the snapshot installation is in progress.

the reason is `installedSnapshotIndex.set(reply.getIndex());` is called in a separate thread in the completeFuture when the snapshot is installed successfully. if we do not wait until the installation is completed, then the following code will lost the chance to be called, and inProgressInstallSnapshotRequest will keep not null and thus `checkInconsistentAppendEntries` will 
always consider the snapshot installation is in progress.
```
      // If a snapshot has been installed, return SNAPSHOT_INSTALLED with the installed snapshot index and reset
      // installedSnapshotIndex to 0.
      long latestInstalledSnapshotIndex = this.installedSnapshotIndex.getAndSet(0);
      if (latestInstalledSnapshotIndex > 0) {
        LOG.info("{}: InstallSnapshot notification result: {}, at index: {}", getMemberId(),
            InstallSnapshotResult.SNAPSHOT_INSTALLED, latestInstalledSnapshotIndex);
        inProgressInstallSnapshotRequest.compareAndSet(firstAvailableLogTermIndex, null);
        return ServerProtoUtils.toInstallSnapshotReplyProto(leaderId, getMemberId(),
            currentTerm, InstallSnapshotResult.SNAPSHOT_INSTALLED, latestInstalledSnapshotIndex);
      }
```  

>So instead of join(), I think we could call wait on the main thread itself for a specified time. This way, if the StateMachine completes the future before the wait if over, we can return the result. Otherwise, we can return IN_PROGRESS notification back to the leader.

if we return IN_PROGRESS notification back to the leader when the wait is over, seems the same problem as described above will happen again, because the future may be completed after the wait . so i thinks there are two possible solutions here.

1 use `join`. this is simple and clear. Although `join` would throw an unchecked exception which might cause the follower to shutdown, a single follower crash will not downgrade the performance of the whole raft group as long as the majority is alive. what is more, the unchecked exception may indicate some serious problems ，such as disk error or network error, so that we have a chance to find and fix the problem and then restart the raft peer.

2 use `get(long timeout, TimeUnit unit)`. it gives us a chance to catch the exception and handle it. but if get  timed out, a timeoutException will be throw out and we should `get` again until the future is completed successfully or exceptionally. what should be done if it is completed exceptionally. if that happens, no more logs can be appended to this raft peer. if leader retry, the same problem will happen again. if so , it is better off shutting it down.

so IMHO, we could use `join` for now, and modify it in the future if need

hanishakoneru commented:
> the reason is installedSnapshotIndex.set(reply.getIndex()); is called in a separate thread in the completeFuture when the snapshot is installed successfully. if we do not wait until the installation is completed, then the following code will lost the chance to be called, and inProgressInstallSnapshotRequest will keep not null and thus checkInconsistentAppendEntries will always consider the snapshot installation is in progress.

The code you mentioned above would be called in the next iteration of the installSnapshot notification. You can verify this from the following statement in the follower log:
```
2021-09-10 14:31:35,854 [grpc-default-executor-0] INFO org.apache.ratis.server.RaftServer$Division: b1b06ffa-12b9-4093-a844-58467d6298b9@group-194BB21401F4: InstallSnapshot notification result: SNAPSHOT_INSTALLED, at index: 2
```
If we send IN_PROGRESS notification to the leader, the leader tries again and in a subsequent attempt, it will get the SNAPSHOT_INSTALLED reply.

We need to figure out why the follower thinks install snapshot is in progress even though it is not. That is what is causing the follower to fail append entries request. From the code path you mentioned above, the LOG statement is being printed in the follower log. So the next line in code `inProgressInstallSnapshotRequest.compareAndSet(firstAvailableLogTermIndex, null);` should also have been run. I am not sure what is happening there.

> 1 use join. this is simple and clear. Although join would throw an unchecked exception which might cause the follower to shutdown, a single follower crash will not downgrade the performance of the whole raft group as long as the majority is alive. what is more, the unchecked exception may indicate some serious problems ，such as disk error or network error, so that we have a chance to find and fix the problem and then restart the raft peer.

I don't think this will be so simple. For example, from Ozone perspective, it would not be desirable for an OzoneManager to crash if there was let's say a network blip or something which caused the install snapshot to fail. Shutting down a peer is an extreme step. We should avoid it especially when the peer is not in an inconsistent state and can be recovered.

JacksonYao287 commented:
@hanishakoneru thanks for your reply!
>We need to figure out why the follower thinks install snapshot is in progress even though it is not

every time the follower receives a `notifyInstallSnapshotRequest` ,  it will get a new object of `firstAvailableLogTermIndex`
```
final TermIndex firstAvailableLogTermIndex = TermIndex.valueOf(
    request.getNotification().getFirstAvailableTermIndex());
```
`inProgressInstallSnapshotRequest` is `AtomicReference<TermIndex>`. when we call `compareAndSet(a, b)`, in fact it will compare the object address(offset in memory) of a and b, but not call something like `compareTo`. so if we reach the following code when leader retry sending `notifyInstallSnapshotRequest` , `firstAvailableLogTermIndex` will not be the one that we set when we receive the first `notifyInstallSnapshotRequest`.so `compareAndSet` will return false and `inProgressInstallSnapshotRequest` will keep not null. and thus `checkInconsistentAppendEntries` will think the installation of the follower is still in progress. 
```
if (latestInstalledSnapshotIndex > 0) {
  LOG.info("{}: InstallSnapshot notification result: {}, at index: {}", getMemberId(),
      InstallSnapshotResult.SNAPSHOT_INSTALLED, latestInstalledSnapshotIndex);
  inProgressInstallSnapshotRequest.compareAndSet(firstAvailableLogTermIndex, null);
  return ServerProtoUtils.toInstallSnapshotReplyProto(leaderId, getMemberId(),
      currentTerm, InstallSnapshotResult.SNAPSHOT_INSTALLED, latestInstalledSnapshotIndex);
}
```
i will fix this in a new commit

>I don't think this will be so simple. For example, from Ozone perspective, it would not be desirable for an OzoneManager to crash if there was let's say a network blip or something which caused the install snapshot to fail. Shutting down a peer is an extreme step. We should avoid it especially when the peer is not in an inconsistent state and can be recovered.

can we use get(long timeout, TimeUnit unit) to wait for a specified duration? what about 10 seconds? if the StateMachine completes the future before the wait is over, we can return the result. Otherwise, we can return IN_PROGRESS notification back to the leader

hanishakoneru commented on ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java:
```diff
@@ -1624,7 +1626,7 @@ private InstallSnapshotReplyProto notifyStateMachineToInstallSnapshot(
         LOG.info("{}: InstallSnapshot notification result: {}", getMemberId(),
             InstallSnapshotResult.SNAPSHOT_UNAVAILABLE);
-        inProgressInstallSnapshotRequest.compareAndSet(firstAvailableLogTermIndex, null);
+        inProgressInstallSnapshotRequest.compareAndSet(firstAvailableLogTermIndex.hashCode(), 0);
```
The hasCode() would still be different between to iterations of this call as we create a firstAvailableLogTermIndex object every time in notifyStateMachineToInstallSnapshot(). 
Instead, we could set inProgressInstallSnapshotRequest to firstAvailableLogIndex. This will be a long value which will not change for different iterations of the same request.

JacksonYao287 replied:
looks good , i will do this in a new commit

hanishakoneru commented on ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java:
```diff
@@ -1604,12 +1604,14 @@ private InstallSnapshotReplyProto notifyStateMachineToInstallSnapshot(
-          throw t;
-        }
+              // wait for 10 seconds for statemachine to install snapshot
+              }).get(10, TimeUnit.SECONDS);
```
10 seconds is too long to block the main thread and wait for this executor thread to finish. I leader could assume this follower is not responding and stop the follower log appended from its side (is there a timeout?). Even if there is no timeout, I dont think we should block the main thread for more than a second.

JacksonYao287 replied:
sure, let us set this to 1 second

hanishakoneru reviewed (commented):
Thanks @JacksonYao287 . This approach looks good to me. 
A few comments posted inline.

JacksonYao287 commented:
@hanishakoneru @szetszwo  i have update this patch according to the comments , please take a look!

hanishakoneru commented:
Thank you @JacksonYao287. Latest patch LGTM. +1.
I will commit this shortly.
